### PR TITLE
os/bluestore: reduce some overhead for _do_clone_range() and _do_remove()

### PR DESCRIPTION
**Pull request by xiexingguo**

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -10393,8 +10393,10 @@ int BlueStore::_do_remove(
           }
         }
 
-	h->extent_map.dirty_range(b_start, b_end - b_start);
-	txc->write_onode(h);
+        if (b_end > b_start) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: I think this should be an assert?  blobs in unshared_blobs are definitely modified and if we don't encounter references to them in the extent_map then there is a bug/problem (we won't dirty and persist the change)

File: src/os/bluestore/BlueStore.cc [hunk 2]
```diff
@@ -10332,74 +10333,82 @@ int BlueStore::_do_remove(
...
+      if (e.logical_offset < b_start) {
+        b_start = e.logical_offset;
+      }
+      if (e.logical_end() > b_end) {
+        b_end = e.logical_end();
       }
     }
   }
+
+  if (b_end > b_start) {
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: this cleanup looks good, but again i think this shoudl be an assert

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (xiexingguo) [inline src/os/bluestore/BlueStore.cc]: @liewegas  okay, updated.

**Comment by xiexingguo** [comment]
this fails test http://pulpito.ceph.com/sage-2017-06-30_05:44:03-rados-wip-sage-testing-distro-basic-smithi/1345006/
will do some local test and verification later

**Comment by liewegas** [comment]
/a/sage-2017-06-30_05:44:03-rados-wip-sage-testing-distro-basic-smithi/1344955 got
<pre>
2017-06-30T06:18:58.257 INFO:tasks.ceph.osd.1.smithi112.stderr:/build/ceph-12.0.3-2457-g4f25a7f/src/os/bluestore/BlueStore.cc: In function 'void BlueStore::ExtentMap::dirty_range(uint32_t, uint32_t)' thread 7f7b36fd6700 time 2017-06-30 06:18:58.260362
2017-06-30T06:18:58.257 INFO:tasks.ceph.osd.1.smithi112.stderr:/build/ceph-12.0.3-2457-g4f25a7f/src/os/bluestore/BlueStore.cc: 2666: FAILED assert(0 == "can't mark unloaded shard dirty")
2017-06-30T06:18:58.258 INFO:tasks.ceph.osd.3.smithi176.stderr:2017-06-30 06:18:58.258698 7f4c9ac68700 -1 received  signal: Hangup from  PID: 30778 task name: /usr/bin/python /usr/bin/daemon-helper kill ceph-osd -f --cluster ceph -i 3  UID: 0
2017-06-30T06:18:58.264 INFO:tasks.ceph.osd.1.smithi112.stderr: ceph version 12.0.3-2457-g4f25a7f (4f25a7f9ca89a19499f31fe81b7d65f70b1dcd74) luminous (rc)
2017-06-30T06:18:58.265 INFO:tasks.ceph.osd.1.smithi112.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x102) [0x5589807ee272]
2017-06-30T06:18:58.265 INFO:tasks.ceph.osd.1.smithi112.stderr: 2: (BlueStore::ExtentMap::dirty_range(unsigned int, unsigned int)+0x54a) [0x558980642bda]
2017-06-30T06:18:58.265 INFO:tasks.ceph.osd.1.smithi112.stderr: 3: (BlueStore::_do_clone_range(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, boost::intrusive_ptr<BlueStore::Onode>&, unsigned long, unsigned long, unsigned long)+0x1298) [0x55898068d448]
2017-06-30T06:18:58.265 INFO:tasks.ceph.osd.1.smithi112.stderr: 4: (BlueStore::_clone_range(BlueStore::TransContext*, boost::intrusive_ptr<BlueStore::Collection>&, boost::intrusive_ptr<BlueStore::Onode>&, boost::intrusive_ptr<BlueStore::Onode>&, unsigned long, unsigned long, unsigned long)+0x288) [0x5589806a7738]
2017-06-30T06:18:58.265 INFO:tasks.ceph.osd.1.smithi112.stderr: 5: (BlueStore::_txc_add_transaction(BlueStore::TransContext*, ObjectStore::Transaction*)+0x1fc2) [0x5589806acfa2]
2017-06-30T06:18:58.266 INFO:tasks.ceph.osd.1.smithi112.stderr: 6: (BlueStore::queue_transactions(ObjectStore::Sequencer*, std::vector<ObjectStore::Transaction, std::allocator<ObjectStore::Transaction> >&, boost::intrusive_ptr<TrackedOp>, ThreadPool::TPHandle*)+0x52e) [0x5589806adbce]
</pre>
need to investigate...

**Comment by liewegas** [comment]
and /a/sage-2017-06-30_05:44:03-rados-wip-sage-testing-distro-basic-smithi/1345006 (ceph-objectstore-tool failure)

**Comment by xiexingguo** [comment]
> need to investigate...

sure, will do.

File: src/os/bluestore/BlueStore.cc [hunk 3]
```diff
@@ -10801,7 +10812,10 @@ int BlueStore::_do_clone_range(
       // make sure it is shared
       if (!blob.is_shared()) {
 	c->make_blob_shared(_assign_blobid(txc), e.blob);
-	dirtied_oldo = true;  // fixme: overkill
+	if (dirty_range_begin == 0) {
+          dirty_range_begin = e.logical_offset;
+        } 
+        dirty_range_end = e.logical_end();
```
Comment (xiexingguo) [inline src/os/bluestore/BlueStore.cc]: should be `dirty_range_end = e.logical_end() - 1;` here to exclude next potential shard

**Comment by xiexingguo** [comment]
> should be dirty_range_end = e.logical_end() - 1; here to exclude next potential shard

Repushed and passed my local ObjectStore/StoreTest, shall be ready for QA again now

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-07-07_18:10:13-rados-wip-sage-testing2-distro-basic-smithi/

**Pull request merged**